### DISCUSSION TITLE
Celery beat + rabbitmq + expiring password

### DISCUSSION THREAD

**Discussion by zenoran** [celery/celery#8429] [General]
I'm trying to integrate celery with a RabbitMQ server which uses Oauth on the backend to auth users.  We have to get a token and use that as the password for logging in to RabbitMQ, but it expires every 60 minutes.  This presents a problem for worker processes which I'm trying to isolate and fix.

Initially we created a custom transport based on pyamqp.SSLTransport which overrides the establish_connection and verify_connection functions to get a new token if it's expired.  This works fine for celery workers, but beat doesn't appear to be calling those same functions prior to publishing messages.

I've attempted to create a custom scheduler based on celery.beat.PersistentScheduler with a modified apply_entry() to call the verify_connection prior to creating the task, but this doesn't appear to be working either.  I can see it's actually executing the verify_connection via the logs, but the password for beat doesn't appear to be using the same one stored in the transport.client (which gets updated via verify_connection).

Does anyone know why beat would be using different cached credentials and if so where I should be updating them?  Also, if I'm going about this all wrong and someone has a better suggestion on how to accomplish this I would greatly appreciate the feedback.

**Comment by dnaeon** [comment]
Having the exact same use case where I need to use OAuth in order to get an access token, which is used for authentication.

@zenoran , did you manage to fix this? Could you please share what you have so far?

Thanks!

**Reply by zenoran** [reply]
It is a literal nightmare trying to change the password once celery is running. I've tried my best to isolate a single point to update but it appears to be stored in many different places in memory.

The best I could come up with was a custom transport for checking the connection, a customized inspect class, and killing beat once the password expires. Not happy with the solution, but I welcome any other advice as well.

**Comment by hernrup-modelon** [comment]
I am trying to do the exact same thing as  you did with the workers (dont need beats) but im having some issues with the events queue. Could you share your solution perhaps? @zenoran

**Comment by rhoban13** [comment]
I'm also very interesting in seeing what this custom `pyamqp.SSLTransport` solution looks like @zenoran.  I suspect I have the same problem with a celery worker using a short lived token for authentication.

**Reply by rhoban13** [reply]
The original post indicated a need to override both `establish_connection` and `verify_connection`.  I'm curious why the latter?  It seems like `establish_connection` gets called upon these disconnects, giving the hook I need to update the transport's copy of the password.
